server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -283,11 +341,157 @@ private void listStartedShards(
                 ActionListener.wrap(listener::onResponse, listener::onFailure)
             );
         }
+
+        Lister<BaseNodesResponse<NodeGatewayStartedShards>, NodeGatewayStartedShards> batchLister = this::batchListStartedShards;
+
+        private synchronized void batchListStartedShards(
+            ShardId shardId,
+            String customDataPath,
+            DiscoveryNode[] nodes,
+            ActionListener<BaseNodesResponse<NodeGatewayStartedShards>> listener
+        ) {
+            pendingFetchShardCount.incrementAndGet();
+            // group shards by node
+            for (DiscoveryNode node : nodes) {
+                Map<ShardId, ShardRequestInfo<NodeGatewayStartedShards>> nodeLevelRequests = queuedRequests.computeIfAbsent(
+                    node,
+                    n -> new HashMap<>()
+                );
+                nodeLevelRequests.put(shardId, new ShardRequestInfo<>(shardId, customDataPath, listener));
```

Comment:
@howardhuanghua  `ShardRequestInfo` need not  to create for each node, the listener in each `ShardRequestInfo` is the same listener.  This can reduce `ShardRequestInfo` Object number from O(shard) * O(node ) to O (shard)

Reply:
Thanks for the comment. I have updated commit to optimize this part.